File: strings/permute.cpp
#include <iostream>

using namespace std;

void permute(const char* instr, char* outstr, bool* used, int len, int level);

int main() {
  const char* const str = "abcd";
  int len = strlen(str);
  char* out = new char[len + 1];
  bool* used = new bool[len]; 

  for (int i = 0; i < len; i++) {
    used[i] = false;
  }

  permute(str, out, used, len, 0);

  return 0;
}

void permute(const char* instr, char* outstr, bool* used, int len, int level) {
  if (level == len) {
    outstr[len] = '\0';
  } else {
    for (int i = 0; i < len; i++) {
      if (!used[i]) {
        outstr[level] = instr[i];
        used[i] = true;
        permute(instr, outstr, used, len, level + 1);
        used[i] = false;
      }
    }
  }
}

File: strings/Permute.java
import java.util.*;
import static java.lang.System.out;

public class Permute {
  public static void main(String... args) {
    String str = (args.length > 0) ? args[0] : "abcd";
    final int strlen = str.length();

    int permutations = factorial(str.length());
    out.println("Number of permutations in input string \"" + str + "\" : " + permutations);

    boolean[] used = new boolean[strlen];
    List<String> permutationStrings = new ArrayList<String>(permutations);
    StringBuilder buffer = new StringBuilder(strlen);

    permute(str, strlen, used, buffer, permutationStrings, 0);

    for (int i = 0; i < permutations; i++) {
      out.println("[" + i + "]=" + permutationStrings.get(i));
    }
  }

  public static int factorial(int n) {
    int f = 1;
    while (n > 1) {
      f *= n--;
    }
    return f;
  }

  public static void permute(final String str, 
                             final int strlen, 
                             final boolean[] used, 
                             final StringBuilder buffer,
                             final List<String> permutationStrings,
                             final int level) {
    assert(level <= strlen);
    assert(used.length == strlen);
    assert(str.length() == strlen);

    // have we completed a permutation?
    if (level == strlen) {
      permutationStrings.add(buffer.toString());
    } else {
      for (int i = 0; i < strlen; i++) {
        if (!used[i]) {
          used[i] = true;
          buffer.append(str.charAt(i));
          permute(str, strlen, used, buffer, permutationStrings,  level + 1);
          buffer.setLength(level);
          used[i] = false;
        }
      }
    }
  }
}

